examples/sudoku_schrodingers_rat.py

```diff
@@ -0,0 +1,347 @@
+
+# emitters 1 = renban, 2 = nabner, 3 = modular, 4 = entropy, 5 = region_sum, 6 = ten_sum
+EMITTERS = {(0, 5): 6, (2, 1): 5, (2, 4): 3, (2, 8): 1, (4, 2): 4, (4, 6): 6, (5, 0): 3, (5, 6): 2, (6, 5): 2, (6, 6): 4, (7, 5): 5, (8, 2): 1}
+
+def gate(idx1, idx2):
+    # the path can not pass from idx2 to idx1, and the value in idx1 must be greater than the value in idx2
+    r1, c1 = idx1
+    r2, c2 = idx2
+    return cp.all([cells[r1,c1] > cells[r2,c2], path[r1,c1] - 1 != path[r2,c2]])
+
+def get_reachable_neighbours(row, column):
+    # from a cell get the indices of its reachable neighbours
+    reachable_neighbours = []
+    if row != 0:
+        if H_WALLS[row-1, column] == 0:
+            reachable_neighbours.append([row - 1, column])
+        if column != 0:
+            if C_WALLS[row-1, column-1] == 0:
+                reachable_neighbours.append([row - 1, column - 1])
+        if column != 8:
+            if C_WALLS[row-1, column] == 0:
+                reachable_neighbours.append([row - 1, column + 1])
+    if row != 8:
+        if H_WALLS[row, column] == 0:
+            reachable_neighbours.append([row + 1, column])
+        if column != 0:
+            if C_WALLS[row, column-1] == 0:
+                reachable_neighbours.append([row + 1, column - 1])
+        if column != 8:
+            if C_WALLS[row, column] == 0:
+                reachable_neighbours.append([row + 1, column + 1])
+    if column != 0:
+        if V_WALLS[row, column-1] == 0:
+            reachable_neighbours.append([row, column - 1])
+    if column != 8:
+        if V_WALLS[row, column] == 0:
+            reachable_neighbours.append([row, column + 1])
+    return reachable_neighbours
+
+def path_valid(path):
+    constraints = []
+    for r in range(path.shape[0]):
+        for c in range(path.shape[0]):
+            neighbours = get_reachable_neighbours(r, c)
+            non_emitter_neighbours = [n for n in neighbours if tuple(n) not in EMITTERS]
+            emitter_neighbours = [n for n in neighbours if tuple(n) in EMITTERS]
+            if (r,c) == (2,8):
+                # The path starts on emitter. It doesn't have any previous cells so the first 3 vectors in the sequence must be fully -1. As for the the last 3, they are taken from the neighbour. Vice versa also applies.
+                constraints.append(cp.all([cp.all(sequence[r,c,:3].flatten() == -1)]))
+                constraints.append(cp.any([cp.all([path[nr,nc] == 2, sequence[r,c,3,0] == cells[nr,nc], sequence[r,c,4,0] == nr, sequence[r,c,5,0] == nc, cp.all(sequence[r,c,3,1:] == sequence[nr,nc,3,:19]), cp.all(sequence[r,c,4,1:] == sequence[nr,nc,4,:19]), cp.all(sequence[r,c,5,1:] == sequence[nr,nc,5,:19]), sequence[nr,nc,0,0] == cells[r,c], sequence[nr,nc,1,0] == r, sequence[nr,nc,2,0] == c, cp.all(sequence[nr,nc,0,1:] == sequence[r,c,0,:19]), cp.all(sequence[nr,nc,1,1:] == sequence[r,c,1,:19]), cp.all(sequence[nr,nc,2,1:] == sequence[r,c,2,:19])]) for nr, nc in neighbours]))
```

Comment:
if these things represent some kind of path, could this path be kept in a separate data structure to make this somewhat more insightfull?

Other way to make this somewhat more readable is to use an auto-formatter like `ruff` on just this files. It will introduce a lot of newlines, but it will be much easier to read and edit.